File: competitions/onlinejudge/ch4/00469/p00469.cpp
#include <iostream>
#include <vector>
#include <sstream>


using namespace std;


struct Coord {
    int r, c;
};

int dr[] = {1, 1, 1, -1, -1, -1, 0, 0};
int dc[] = {-1, 0, 1, -1, 0, 1, -1, 1};



int floodfill(vector<vector<char>> &grid, int r, int c, char c1, char c2) {
    int ans = 0;
    int rows = grid.size();
    int columns = grid[0].size();

    if (r < 0 || r > rows - 1 || c < 0 || c > columns - 1) return 0;
    if (grid[r][c] != c1) return 0;

    ans = 1;

    grid[r][c] = c2;

    for(int d = 0; d<8; d++) {
        ans += floodfill(grid, dr[d] + r, dc[d] + c, c1, c2);
    }

    return ans;
}


int main() {
    int n;
    string s;
    stringstream ss;
    Coord c;

    cin>>n;
    getline(cin, s);
    getline(cin, s);

    for(int i = 0; i < n; i++) {
        vector<vector<char>> grid;
        vector<Coord> coords;
        while (getline(cin, s)) {
            if (s == "") break;

            if (s[0] == 'L' || s[0] == 'W') {
                grid.push_back({});
                for (int j = 0; j<s.length(); j++) {
                    grid[grid.size() - 1].push_back(s[j]);
                }
            } else {
                ss << s;
                ss >> c.r >> c.c;
                ss.str("");
                ss.clear();

                c.r = c.r - 1;
                c.c  = c.c - 1;
                coords.push_back(c);
            }
        }
        
        for(auto c: coords) {
            vector<vector<char>> new_grid = grid;
            cout<<floodfill(new_grid, c.r, c.c, 'W', '.')<<endl;
        }
        
        if(i == n -1) continue;
        cout<<endl;
    } 

    return 0;
}

File: competitions/onlinejudge/ch4/00469/p00469.py
import sys
from copy import deepcopy


stdin = sys.stdin


dr = [1, 1, 1, -1, -1, -1, 0, 0]
dc = [-1, 0, 1, -1, 0, 1, -1, 1]


def floodfill(grid, r, c, c1, c2):
    ans = 0
    rows, columns = len(grid), len(grid[0])
    if r < 0 or r > rows - 1 or c < 0 or c > columns - 1:
        return 0

    if grid[r][c] != c1:
        return 0

    ans = 1
    grid[r][c] = c2
    for i in range(8):
        ans += floodfill(grid, r + dc[i], c + dr[i], c1, c2) 

    return ans


if __name__ == '__main__':
    n = int(stdin.readline().strip())
    stdin.readline()

    for i in range(n):
        grid = []
        coords = []
        for row in stdin:
            row = row.strip()
            if row[0].isnumeric():
                break
            grid.append(list(row))

        coords.append([int(x) for x in row.split()])
        for row in stdin:
            row = row.strip()
            if not row:
                break
            coords.append([int(x) for x in row.split()])

        for r, c in coords:
            print(floodfill(deepcopy(grid), r - 1, c - 1, 'W', '.'))
        if i == n - 1:
            continue
        print()
